### ISSUE TITLE
Adding multiple modules with a single backing state

### ISSUE THREAD

**Issue by jpreese** [gruntwork-io/terragrunt#781]
If I have a generic enough module where I can spin up multiple of them with just changing my inputs, how would I implement that in Terragrunt? Every example I'm seeing points to a single source, and has inputs against a single module.

I've tried putting a root level terragrunt.hcl in a folder, and then two subfolders with terragrunt.hcl files representing each of the modules and their inputs, so they both get the same remote state, but I'm getting some state locking errors.

Presumably because it's trying to apply the modules in parallel to the same state file, which is my intention. So either I need a way to unparallelize the plan/apply -- or I am going about this wrong.

TL;DR --

Need to apply 2+ modules that only differ by input values that are backed by a a single remote state key.

In terraform I'd just use `module { }` twice, but there doesn't appear to be any similar examples with Terragrunt.

**Comment by ekini** [comment]
What remote state backend do you use? Why does it have to be the same state file/key?

I have quite a few generic modules and use them several times. The state backend is configured as follows:
```hcl
remote_state {
  backend = "s3"

  config = {
    bucket  = "infrastructure-xxxxxxx"
    key     = "terraform/${path_relative_to_include()}/terraform.tfstate"
...
  }
}
```
so that every terragrunt module gets its own state file.

**Comment by jpreese** [comment]
It's azurerm.

The issue is mostly around if you have a very small module, lets say a module that stands up a vnet and a subnet. The only difference is some inputs, maybe a name and some ip ranges. But you want to create multiple vnets using the same state file.

```
network/
terragrunt.hcl
-- useast
   -- terragrunt.hcl (references small module, with some inputs)
-- uswest
   -- terragrunt.hcl (references same small module, with different inputs)
```

The above would create two distinct state files, which seems like overkill the more and more you grow.

**Comment by brikis98** [comment]
> Presumably because it's trying to apply the modules in parallel to the same state file, which is my intention.

Why? Perhaps I'm misunderstanding you, but this sounds like the exact opposite of how Terraform state should be used. See [How to manage Terraforms state](https://blog.gruntwork.io/how-to-manage-terraform-state-28f5697e68fa).

**Comment by jpreese** [comment]
I don't necessarily want to perform the operations in parallel, what I'm getting stuck on is how to use the same module, with different inputs, but have the result be added to the same state key.

In "native" terraform I would just do this:

```hcl
  module "vnet_east" {
    source = "../../../modules/azure/vnet"

    name               = "east-vnet"
    address_space_cidr = "10.10.0.0/22"
  }

  module "vnet_west" {
    source = "../../../modules/azure/vnet"

    name               = "west-vnet"
    address_space_cidr = "10.11.0.0/22"
  }

.. do stuff with the module outputs ..

```

In terragrunt, we use `source =` and `inputs {` which performs the operation in parallel with differing keys.

**Comment by jpreese** [comment]
Actually it looks like this request is in proposal https://github.com/gruntwork-io/terragrunt/issues/759

**Comment by brikis98** [comment]
Create a Terraform module (`.tf` files) exactly as you showed above, with multiple `module "..." { ... }` blocks in it, and run `terragrunt apply` on that module to get what you want.

**Comment by so0k** [comment]
trying the above gives me errors with relative paths ... the composite module can not refer to the basic modules it uses once terragrunt copied it over into the temporary terragrunt cache directory

my folder layout is as follows (this is a first iteration and we have not split modules of into their own repo yet)

```
my-repo
$ tree -L 3
.
├── README.md
├── live
│   ├── management
│   │   ├── network-foo < uses network-type1
│   │   └── terragrunt.hcl
│   ├── non-prod
│   │   ├── bar
│   │   └── baz
│   └── prod
│       └── qux
└── modules
    ├── basic
    │   ├── vpc
    │   ├── private-subnets
    │   └── public-subnets.tf
    └── composed
        ├── network-type1
        └── main.tf <refers to ../basic/vpc,private-subnets,public-subnets
```

the composed module defines the providers block re-used by the basic modules as well as an empty backend configuration... (which are not in the basic modules)

**Comment by yorinasub17** [comment]
See https://github.com/gruntwork-io/terragrunt#remote-terraform-configurations, specifically the following note:

```
(Note: the double slash (//) in the source parameter is intentional and required. It's part of Terraform's Git syntax for module sources. Terraform may display a "Terraform initialized in an empty directory" warning, but you can safely ignore it.)
```

**Comment by so0k** [comment]
but I'm using local modules... how does the double path apply there? @yorinasub17

**Comment by so0k** [comment]
terraform/go-getter doesn't support Local source with double slash ...

https://www.terraform.io/docs/modules/sources.html#local-paths

**Comment by yorinasub17** [comment]
The double slash still applies for local modules and **Terragrunt** supports it. It is the only way to know what to use as the root.

See the note at https://github.com/gruntwork-io/terragrunt#working-locally

> (Note: the double slash (//) here too is intentional and required. Terragrunt downloads all the code in the folder before the double-slash into the temporary folder so that relative paths between modules work correctly. Terraform may display a "Terraform initialized in an empty directory" warning, but you can safely ignore it.)

**Comment by so0k** [comment]
oh thanks, it works now!

**Comment by yorinasub17** [comment]
Closing due to inactivity. Please open new issues if you still have problems.

**Comment by adi4dpeople** [comment]
I couldn't find a more suitable issue to address my problem than this so I'm reopening this. I have a setup where i have environments folder living inside my modules directory and it structure is like below. Only the values inside .tfvars vary per environment but remaining all are duplicated. So I'm testing terragrunt currently. But, I couldn't find terragrunt documentation on how to create multiple resources from same source module.  If i start writing modules similar to what you have suggested to @jpreese, i would still end up duplicating the code because i will have a main.tf in which i call all the modules, different values for each modules so i need have vars.tf and finally i need to pass values through tfvars. So in my scenario how can I DRY my code?   

current environments directory structure. 
```
├── environments
│   ├── hub
│   │   ├── datasources.tf
│   │   ├── extract_tf_vars.py
│   │   ├── main.tf
│   │   ├── prod_hub.tfvars
│   │   ├── provider.tf
│   │   └── vars.tf
│   ├── perf
│   │   ├── datasources.tf
│   │   ├── extract_tf_vars.py
│   │   ├── main.tf
│   │   ├── perf.tfvars
│   │   ├── provider.tf
│   │   └── vars.tf
│   ├── prod
│   │   ├── datasources.tf
│   │   ├── extract_tf_vars.py
│   │   ├── main.tf
│   │   ├── prod.tfvars
│   │   ├── provider.tf
│   │   └── vars.tf
│   └── trng
│       ├── datasources.tf
│       ├── extract_tf_vars.py
│       ├── main.tf
│       ├── provider.tf
│       ├── trng.tfvars
│       └── vars.tf
```

Right now this is how I'm calling the modules from terragrunt. 
```
aditya@LAPTOP-6C2MPJDV:~$ cat terraform_live/perf/virtual_network/terragrunt.hcl
terraform {
  source = "git::git@github.com:adi4dpeople/terraform_modules.git//virtual_network?ref=v0.0.1"
}

# Include all settings from the root terragrunt.hcl file
include {
  path = find_in_parent_folders()
}

dependency "resource_group" {
  config_path = "../resource_group"
}

# These are the variables we have to pass in to use the module specified in the terragrunt configuration above
inputs = {
  vnet_name = "testvnet"
  vnet_address_space = "1.2.3.4/16"
  resource_group_name = dependency.resource_group.outputs.name
}
```

**Comment by yorinasub17** [comment]
If I understand what you want correctly, your issue is different from this particular issue as this issue is looking to consolidate multiple modules into a single state file. Will answer in the other ticket.

**Comment by dtcMLOps** [comment]
Hi folks, is there any update regarding this approach?

My use case is with Databricks since we are creating multiple Databricks workflows per project/country-project/environment and for me would be better to handle 1 state file with all the infra from 1 environment instead of having multiple state files per sub-folder (country or project).

The issue I have found with this is if we remove 1 country or project, the state file for that removed project will remain stored in the remote backend.

**Comment by nikdvy** [comment]
my case: we use 44 dynamodb tables with tmpfs (in-memory fs) for gitlab runner. it takes 660Mi per 1 dynamodb table as it downloads the aws provider per installation. then i end up with 28Gi memory and need in 32Gi ec2 instance in aws for this particular pipeline which is weird. why do i need to download the same provider 44 times?

**Comment by yhakbar** [comment]
@nikdvy you don't!

I recommend taking a look at the following and seeing if they help you:
- https://terragrunt.gruntwork.io/docs/features/auto-provider-cache-dir/
- https://terragrunt.gruntwork.io/docs/features/cas/
- https://terragrunt.gruntwork.io/docs/features/provider-cache-server/
